storage/remote/codec.go

```diff
@@ -183,6 +183,150 @@ func FromQueryResult(res *prompb.QueryResult) storage.SeriesSet {
 	}
 }
 
+func StreamChunkedReadResponses(
+	stream io.Writer,
+	queryIndex int64,
+	ss storage.SeriesSet,
+	sortedExternalLabels []prompb.Label,
+	sampleLimit int,
+	maxChunksInFrame int,
+) error {
+	var numSamples = 0
+
+	for ss.Next() {
+		series := ss.At()
+		iter := series.Iterator()
+		lbls := MergeLabels(labelsToLabelsProto(series.Labels()), sortedExternalLabels)
+
+		for iter.Next() {
+			// TODO(bwplotka): Expose maxChunks as flag?
+			chks, samples, err := encodeChunks(iter, 120, maxChunksInFrame, sampleLimit-numSamples)
+			if err != nil {
+				return err
+			}
+			numSamples += samples
+			if len(chks) == 0 {
+				break
+			}
+
+			b, err := proto.Marshal(&prompb.ChunkedReadResponse{
+				// TODO(bwplotka): Do we really need multiple?
+				ChunkedSeries: []*prompb.ChunkedSeries{
+					{
+						// TODO(bwplotka): Oneof? We don't need to send labels all the time.
+						Labels: lbls,
+						Chunks: chks,
+					},
+				},
+				QueryIndex: queryIndex,
+			})
+			if err != nil {
+				return errors.Wrap(err, "marshal ChunkedReadResponse")
+			}
+
+			if _, err := stream.Write(b); err != nil {
+				return errors.Wrap(err, "write to stream")
+			}
+		}
+
+		if err := iter.Err(); err != nil {
+			return err
+		}
+	}
+	if err := ss.Err(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// encodeChunks expects iterator to be ready to use (aka iter.Next() done before invoking).
+func encodeChunks(iter storage.SeriesIterator, maxSamplesPerChunk int, maxChunks int, allowedNumSamples int) ([]prompb.Chunk, int, error) {
+	var (
+		chkMint int64
+		chk     *chunkenc.XORChunk
+		app     chunkenc.Appender
+		err     error
+
+		isNext     = true
+		numSamples = 0
+		chks       = make([]prompb.Chunk, 0, maxChunks)
+	)
+
+	for isNext {
+		numSamples++
+		if allowedNumSamples > 0 && numSamples > allowedNumSamples {
+			return nil, numSamples, HTTPError{
+				msg:    fmt.Sprintf("exceeded sample limit"),
+				status: http.StatusBadRequest,
+			}
+		}
+
```

Comment:
Do we need a sample limit in the streaming case?

Reply:
Good question, what are your thoughts? (: If you accidently ask for 10 years data you will just.. timeout I guess? so timeout might be totally enough.

Reply:
Happy to remove sample limitation

Reply:
Removed, and added info in sample limit flag.